pkg/network/ovn_kubernetes.go

```diff
@@ -1522,21 +1822,51 @@ func shouldUpdateOVNKonPrepull(ovn bootstrap.OVNBootstrapResult, releaseVersion
 }
 
 // shouldUpdateOVNKonUpgrade determines if we should roll out changes to
-// the master and node daemonsets on upgrades. We roll out nodes first,
-// then masters. Downgrades, we do the opposite.
-func shouldUpdateOVNKonUpgrade(ovn bootstrap.OVNBootstrapResult, releaseVersion string) (updateNode, updateMaster bool) {
+// the master and node daemonsets and control plane deployment (according to the cluster zone mode) upon upgrades.
+// We roll out node first, then master and/or control plane. In downgrades, we do the opposite.
+//
+// If we're upgrading from a 4.13 cluster, which has no OVN InterConnect support, three phases are necessary.
+// Phase 1:
+//
+//	a) prepareUpgradeToInterConnect pushes a configMap with zone-mode=singlezone, temporary=true;
+//	b) renderOVNKubernetes selects the YAMLs from the single-zone folder
+//	   (node DS, master DS [SS for hypershift], ovnkube-sbdb route [hypershift]);
+//	c) shouldUpdateOVNKonUpgrade rolls out first node DS then master DS/SS in single-zone mode
+//	   (there's no zone-mode change, since the 4.13 architecture is equivalent to single zone).
+//
+// Phase 2tmp:
+//
+//	a) Master and Node DS/SS are now single zone, so prepareUpgradeToInterConnect overrides the configMap
+//	   with zone-mode=multizone, temporary=false;
+//	b) renderOVNKubernetes selects the YAMLs from the multi-zone-tmp folder (node DS, master DS/SS,
+//	   control plane deployment, ovnkube-sbdb route [hypershift]);
+//	c) shouldUpdateOVNKonInterConnectZoneModeChange applies a zone mode change (single->multi) by rolling out
+//	   first node DS and then master DS/SS+control plane deployment.
+//
+// Phase 2:
+//
+//	a) node DS is multizone, control plane is multizone, but we still have singlezone master DS/SS;
+//	b) renderOVNKubernetes selects the YAMLs from the multi-zone folder (node DS, master DS), effectively
+//	   removing the old ovnk master DS/SS and, in hypershift, the old ovnkube-sbdb route;
+//	c) prepareUpgradeToInterConnect removes the configMap.
+//
+// At the end, we have a 4.14 cluster in multizone mode.
+//
+// 4.15 CNO won't need this extra complexity, since only multizone will be supported.
+// TODO: in 4.15, no need to keep this logic, once we've required to upgrade to 4.15 only from 4.14.
```

Comment:
I wanted to document in one place how the upgade to IC happens. Should it be in a separate file perhaps? I simply chose `shouldUpdateOVNKonUpgrade` because it's where the upgrade starts, but yeah, it could have been elsewhere as well.
Rolling out node before master is already what we do today:
https://redhat-internal.slack.com/archives/C050MC61LVA/p1689945804528209?thread_ts=1689928536.699009&cid=C050MC61LVA

In 4.15 we can get rid of the single-zone folder and the multizone-tmp folder and the whole upgrade path from non-IC to IC.  As for the ordering between node and control-plane, this can be rethought as well in 4.15, since I don't see any dependency between cluster-manager and ovnk node. For simplicity in this PR I'm swapping the role of the old ovnk master with the new ovnk control plane when we're switching to multizone, in order to keep compatibility with the existing code, and with the possibility of switching back to single-zone, but once we're in 4.15 this can be further simplified.

Reply:
Ok, I thought it was a bit misleading. Perhaps the bet place would be `prepareUpgradeToInterConnect`

Reply:
Sure!